### ISSUE TITLE
packet buffer headers should have protocol number associated.

### ISSUE THREAD

**Issue by jelischer** [google/gvisor#3810]
Packet headers are now tracked in the packet buffer structure using a special 'header' structure. This structure should maintain what type each header is as well as what layer it is (which is all it has now). In parts of the code where it is important to look at the transport header for example it is expensive to go back to re-examine the IPv6 header to find which kind of transport header is in the transport header storage (tcp? udp? icmp?). Since we already know this information at some stage  it is probably worth while to save it with the header.

**Comment by jelischer** [comment]
It turns out that to do this we probably should fix the fact that ICMP never does a Consume() operation on the packet for the transport header, but leaves it in the Data segment. The reason for this is that ICMP processing happens even if the ICMP protocol is not available for use by upper layers and the setting of the header into its own section would normally be done by the upper ICMP code. The answer would be to have it done at the lower ICMP code in network/ipv[46]/icmp.c which is always present.

**Comment by tamird** [comment]
cc @kkrakauer

**Comment by kevinGC** [comment]
Yes, there is a confusing coupling between ICMP _sockets_ (enabled via `stack.Options.TransportProtocols`) and ICMP handling. They should share code, but the latter shouldn't depend on enabling the former.

For context: this came up when we added early header parsing, but wasn't needed immediately and so nobody allocated cycles for it.

**Comment by jelischer** [comment]
I could add some cycles to it while I have it  fully loaded in my head.

**Comment by ghanan94** [comment]
This is done - packet buffers hold the network/transport protocol numbers

**Comment by tamird** [comment]
There are still 5 TODOs referencing this issue.
